app/src/main/java/org/grammatek/simacorrect/spellcheckerservice/YfirlesturAnnotation.kt

```diff
@@ -35,105 +40,109 @@ class YfirlesturAnnotation(
      * For storing the starting index, end index and length for all annotations.
      * Used by the spell checker service to determine where to place annotations.
      */
-    class SuggestionIndexes(
+    class AnnotationIndices (
         val startChar: Int,
-        endChar: Int,
+        val endChar: Int,
     ) {
         var length: Int = endChar - startChar + 1
     }
 
     /**
      * Creates a [Key] data class from the
-     * annotation.start and annotation.end indexes.
+     * annotation.start and annotation.end indices.
      * Used to identify which annotations belong together.
      */
     private fun Annotations.toKey() = Key(start, end)
     data class Key(val startIndex: Int?, val endIndex: Int?)
 
     /**
      * Iterates through all the [_annotations] and builds
-     * a [SuggestionsInfo] using data in the [_annotations] as
+     * a list of [SuggestionsInfo] using the data in the [_annotations] as
      * well as determining the type of spelling error.
      *
      * @returns List<SuggestionsInfo>
      */
-    fun getSuggestionsForAnnotatedWords(suggestionsLimit: Int): List<SuggestionsInfo> {
-        // Group annotations in a list that have the same start AND end index.
-        // Necessary to distinguish between cases where single word annotations
-        // are inside of multi word annotations.
-        val annotationsList = _annotations.groupBy { it.toKey() }
+    fun getSuggestionsForAnnotatedWords(suggestionsLimit: Int, dict: ArrayList<String>): List<SuggestionsInfo> {
         val suggestionList = mutableListOf<SuggestionsInfo>()
+        for (i in _annotations.indices) {
+            // Group annotations in a list that have the same start AND end index.
+            // Necessary to distinguish between cases where single word annotations
+            // are inside of multi word annotations.
```

Comment:
The whole function is still very hard to understand, could you add some short text examples ?

Reply:
I was able to refactor it quite much after removing the method `adjustTokenIndices`. Also added some comments, please take another look.